pkg/network/tracer/ebpf_conntracker.go

```diff
@@ -0,0 +1,330 @@
+			}
+		}
+	}
+}
+
+func (e *ebpfConntracker) addTranslation(src *ConnTuple, dst *ConnTuple) error {
+	if err := e.ctMap.Update(unsafe.Pointer(src), unsafe.Pointer(dst), ebpf.UpdateNoExist); err != nil && !errors.Is(err, ebpf.ErrKeyExist) {
+		return err
+	}
+	return nil
+}
+
+func formatKey(netns uint32, tuple *ct.IPTuple) *ConnTuple {
+	var proto network.ConnectionType
+	switch *tuple.Proto.Number {
+	case unix.IPPROTO_TCP:
+		proto = network.TCP
+	case unix.IPPROTO_UDP:
+		proto = network.UDP
+	default:
+		return nil
+	}
+
+	return newConnTuple(0,
+		netns,
+		util.AddressFromNetIP(*tuple.Src),
+		util.AddressFromNetIP(*tuple.Dst),
+		*tuple.Proto.SrcPort,
+		*tuple.Proto.DstPort,
+		proto)
+}
+
+func (e *ebpfConntracker) GetTranslationForConn(stats network.ConnectionStats) *network.IPTranslation {
+	start := time.Now()
+	src := tuplePool.Get().(*ConnTuple)
+	defer tuplePool.Put(src)
+
+	if err := toConnTupleFromConnectionStats(src, &stats); err != nil {
+		return nil
+	}
+	src.pid = 0
+	log.Tracef("looking up in conntrack: %s", src)
+
+	dst := e.get(src)
+	if dst == nil {
+		return nil
+	}
+	defer tuplePool.Put(dst)
+
+	atomic.AddInt64(&e.stats.gets, 1)
+	atomic.AddInt64(&e.stats.getTotalTime, time.Now().Sub(start).Nanoseconds())
+	return &network.IPTranslation{
+		ReplSrcIP:   dst.SourceAddress(),
+		ReplDstIP:   dst.DestAddress(),
+		ReplSrcPort: uint16(dst.sport),
+		ReplDstPort: uint16(dst.dport),
+	}
+}
+
+func (e *ebpfConntracker) get(src *ConnTuple) *ConnTuple {
+	dst := tuplePool.Get().(*ConnTuple)
+	if err := e.ctMap.Lookup(unsafe.Pointer(src), unsafe.Pointer(dst)); err != nil {
+		if !errors.Is(err, ebpf.ErrKeyNotExist) {
+			log.Warnf("error looking up connection in ebpf conntrack map: %s", err)
+		}
+		tuplePool.Put(dst)
+		return nil
+	}
+	return dst
+}
+
+func (e *ebpfConntracker) delete(key *ConnTuple) {
+	if err := e.ctMap.Delete(unsafe.Pointer(key)); err != nil {
+		if errors.Is(err, ebpf.ErrKeyNotExist) {
+			log.Tracef("connection does not exist in ebpf conntrack map: %s", key)
+			return
+		}
+		log.Warnf("unable to delete conntrack entry from eBPF map: %s", err)
+	}
+}
+
+func (e *ebpfConntracker) DeleteTranslation(stats network.ConnectionStats) {
+	start := time.Now()
+	key := tuplePool.Get().(*ConnTuple)
+	defer tuplePool.Put(key)
+
+	if err := toConnTupleFromConnectionStats(key, &stats); err != nil {
+		return
+	}
+	key.pid = 0
+
+	dst := e.get(key)
+	e.delete(key)
+	if dst != nil {
+		e.delete(dst)
+		tuplePool.Put(dst)
+	}
+	atomic.AddInt64(&e.stats.unregisters, 1)
+	atomic.AddInt64(&e.stats.unregistersTotalTime, time.Now().Sub(start).Nanoseconds())
+}
+
+func (e *ebpfConntracker) GetStats() map[string]int64 {
+	m := map[string]int64{
+		"state_size": 0,
+	}
+	telemetry := &conntrackTelemetry{}
+	if err := e.telemetryMap.Lookup(unsafe.Pointer(&zero), unsafe.Pointer(telemetry)); err != nil {
+		log.Tracef("error retrieving the telemetry struct: %s", err)
+	} else {
+		m["registers_total"] = int64(telemetry.registers)
```

Comment:
Worth adding string consts for these?

Reply:
Perhaps, but this PR is large enough as-is. I'm going to leave that refactoring for a future PR.